wartool.cpp

```diff
@@ -3036,25 +3079,135 @@ int main(int argc, char** argv)
 					copyArchive(copyfile);
 				}
 				sprintf(extract, "%s/%s", Dir, Todo[u].File);
-				if (Todo[u].Arg2 == 1) { // compress
-					sprintf(extract, "%s.gz", extract);
-				}
-				if (Todo[u].Arg2 == 2) { // video file
-					sprintf(extract, "%s.smk", extract);
-				}
-				if (Todo[u].Arg2 == 8 && !rip) { // wav audio
-					continue;
+				switch (Todo[u].Arg2) {
+					case 1:
+						sprintf(extract, "%s.gz", extract);
+						break;
+					case 'V':
+						sprintf(extract, "%s.smk", extract);
+						break;
+					case 'W':
+						if (!rip) { // wav audio
+							continue;
+						}
+					default:
+						break;
 				}
 				if (ExtractMPQFile(mpqfile, (char*)Todo[u].ArcFile, extract, Todo[u].Arg2 == 1)) {
 					printf("Failed to extract \"%s\"\n", (char*)Todo[u].ArcFile);
 				}
-				if (Todo[u].Arg2 == 2) { // convert videos
-					if (video) {
-						ConvertVideo(Todo[u].File, Todo[u].Arg1, true);
-					}
-				}
-				if (Todo[u].Arg2 == 4) { // convert videos
-					ConvertPud(Todo[u].File, 0, true);
+				switch (Todo[u].Arg2) { // same mapping as outer loop
+					case 'S':
+						{
+							std::ifstream f(extract, std::ios::in|std::ios::out|std::ios::binary);
+							f.seekg(0, std::ios::end);
+							int sz = f.tellg();
+							f.seekg(0, std::ios::beg);
+							char *buf = new char[sz];
+							char *currentBuf = buf + Todo[u].Arg3;
+							f.read(buf, sz);
+							UnitNamesLast = 0;
+							while (UnitNamesLast < (sizeof(UnitNames) / sizeof(*UnitNames))) {
+#ifdef WIN32
+								UnitNames[UnitNamesLast] = _strdup(currentBuf);
+#else
+								UnitNames[UnitNamesLast] = strdup(currentBuf);
+#endif
+								currentBuf += strlen(currentBuf) + 1;
+								if (currentBuf - buf >= sz) {
+									break;
+								}
+								UnitNamesLast++;
+							}
+							delete[] buf;
+						}
+						break;
+					case 'V':
+						if (video) {
+							ConvertVideo(Todo[u].File, Todo[u].Arg1, true);
+						}
+						break;
+					case 'P':
+						ConvertPud(Todo[u].File, 0, true);
+						break;
+				    case 'X':
+						{
+							std::fstream f(extract, std::ios::in|std::ios::out|std::ios::binary);
+							f.seekg(0, std::ios::end);
+							int sz = f.tellg();
+							f.seekg(0, std::ios::beg);
+							unsigned char *buf = new unsigned char[sz];
+							f.read((char*)buf, sz);
+							f.seekg(0, std::ios::beg);
+							unsigned char *str;
+							str = ConvertString(buf + Todo[u].Arg3, sz - Todo[u].Arg3);
+							f.write((char*)str, strlen((char*)str));
```

Comment:
Looking at the end of a text file, you will see a few extra characters. It is because you ignore the `Todo[u].Arg3` n char, but without truncating the file when you wrote it. Because you write over the existing file, you will have the "old" char in the stream. Quickfix : add `fwrite("\0", 1)` just after writing the str.